int64 procedure
string state
string error